Cover: Use context to show Upgrade Nudge.

retrofox opened the pull request:
As part of the "Next Steps" of https://github.com/Automattic/jetpack/pull/16107, this PR refactorizes the code of the shared/blocks/cover block to:

1) Stop using a constant for the allowed files. Instead of this, replace it with the `allowedMimeTypes` object defined in the Jetpack initial state. It's handled by the `isVideoFile` helper function.
2) Add the `isUpgradable()` helper to make the code more readable and reusable.
3) Handle data propagation between CoverEditBlock and the media placeholder through of a new context.

#### Changes proposed in this Pull Request:
* Connect with context. Add `isVideoFile()` and `isUPgradable()` helper functions.

#### Testing instructions:

* Apply D44997-code
* get sandboxed
* Confirm that when trying to upload a video for a cover block using the MediaPlaceholder will show the Upgrade Nudge, only for Simple sites with a free plan. Otherwise, it shouldn't show the Nudge.

![image](https://user-images.githubusercontent.com/77539/84763133-1def3480-afa2-11ea-99d5-5168e1a99af8.png)


#### Proposed changelog entry for your changes:
* Connect with context. Add `isVideoFile()` and `isUPgradable()` helper functions.

getdave commented on extensions/shared/blocks/cover/utils.js:
```diff
@@ -1,19 +1,59 @@
+		return fileExtension && allowedVideoFileExtensions.includes( fileExtension );
+	}
+
+	allowedVideoMimeTypes = values( allowedVideoMimeTypes );
```
This appears to be divorced from the declaration of `let allowedVideoMimeTypes`. Any reason we don't just do this as a single operation and use `const`? Also why are they no collocated in the code?

retrofox replied:
yes, the reason is that when I create `allowedVideoFileExtensions` constant I need the data collection. Once I get it, then and cover `allowedVideoMimeTypes` to an Array of string items.

retrofox commented:
> I tested this on Simple using the patch. I found that it works ok but there is one "glitch" that I feel should be solved.

Yes, that's true but it's something that wasn't introduced on the PR. You can reproduce a similar behavior with the `video` block. Provably something to be fixed but in another PR imo.


> Also, I noticed a less ideal piece of UX in this block is that there is no feedback when uploading an image.

Definitely `<MediaPlaceholder />` needs more ❤️, but the same: It's our of the scope of this PR. :-D

> I assume this is unrelated and should be addressed in Gutenberg Core?

Yes, we also have all resources to handle on this side, though.

@getdave thanks for the review. I consider that these issues should be tackled in another PR since they don't belong to the scope of this PR. :-D

getdave commented:
> Yes, that's true but it's something that wasn't introduced on the PR. You can reproduce a similar behavior with the video block. Provably something to be fixed but in another PR imo.

Done https://github.com/Automattic/jetpack/issues/16174.

Also [raised an Issue in Gutenberg Core](https://github.com/WordPress/gutenberg/issues/23206) to track the issue with Cover block providing no feedback when uploading.

> @getdave thanks for the review. I consider that these issues should be tackled in another PR since they don't belong to the scope of this PR. :-D

Yep that's fine. As indicated, I wanted to check these weren't directly related. As they aren't then this is good to go.

obenland commented on extensions/shared/blocks/cover/utils.js:
```diff
@@ -1,19 +1,59 @@
+ * Internal dependencies
+ */
+import { isSimpleSite } from "../../site-type-utils";
+import getJetpackExtensionAvailability from "../../get-jetpack-extension-availability";
```
I'm surprised linting didn't flag the double quote usage :)

retrofox replied:
It's because your 👀 :-)
I haven't ignored the linting btw. Updating

obenland commented on extensions/shared/blocks/cover/utils.js:
```diff
@@ -1,19 +1,59 @@
+	}
+
+	let allowedVideoMimeTypes = pickBy( allowedMimeTypes, ( type ) => /^video\//.test( type ) );
+	const allowedVideoFileExtensions = flatten( map( keys( allowedVideoMimeTypes ), ext => ext.split( '|' ) ) );
```
I noticed `ext` is passed without parens, while in the line above `type` is passed with parens.
Not sure which version is preferred, but let's stay consistent.

obenland commented on extensions/shared/blocks/cover/utils.js:
```diff
@@ -1,19 +1,59 @@
+		return false;
+	}
+
+	if ( ! file ) {
```
I wonder if this check should be at the top of the function?
Alternatively it could also be combined with the one above.

obenland commented:
Testing with the Maywood theme on wp.com, the notice text was white on a white background and illegible. We might need to add some CSS to account for varying font colors.

retrofox commented on extensions/shared/blocks/cover/utils.js:
```diff
@@ -1,19 +1,59 @@
+	const allowedVideoFileExtensions = flatten( map( keys( allowedVideoMimeTypes ), ext => ext.split( '|' ) ) );
+
+	if ( typeof file === 'string' ) {
+		const fileExtension = ( file.split( '.' ) )?.[ 1 ];
```
Good catch! and even a better solution. Thanks!

retrofox commented:
> Testing with the Maywood theme on wp.com, the notice text was white on a white background and illegible. We might need to add some CSS to account for varying font colors.

Good catch. Working on a solution.

retrofox commented:
> Testing with the Maywood theme on wp.com, the notice text was white on a white background and illegible. We might need to add some CSS to account for varying font colors.

I've created a context in order to move the Nudge element to another place, out of the CoverEditBlock component. It grew up ;-).

obenland commented:
It looks like D44997-code is out of sync and needs a refresh

retrofox commented:
> It looks like D44997-code is out of sync and needs a refresh

refreshed!

obenland commented:
Thank you!

Testing on my sandbox, the upgrade banner appears behind the toolbar, I'm not sure that's intentional?
![Screen Shot 2020-06-17 at 4 01 43 PM](https://user-images.githubusercontent.com/1398304/84959479-42d3cc80-b0b4-11ea-9be6-117b1886576f.png)


Does this PR intend to also fix the "Replace" flow, or is that separate? This is what I get if I want to replace a cover image with a video:
![Screen Shot 2020-06-17 at 4 02 40 PM](https://user-images.githubusercontent.com/1398304/84959546-66971280-b0b4-11ea-8fd5-42a1fd2faa66.png)

retrofox commented:
> Does this PR intend to also fix the "Replace" flow, or is that separate? 

No, it doesn't. It's going to be tackled in a separate one.

retrofox commented:
> Testing on my sandbox, the upgrade banner appears behind the toolbar, I'm not sure that's intentional?

I think the toolbar always will be at the front of all components. What we should do is try to find a way to close it once the Nudge appears. Maybe in a following PR?

obenland commented:
[In Dave's comment](https://github.com/Automattic/jetpack/pull/16171#pullrequestreview-432384215) it appears between the toolbar and placeholder. Did that change?

retrofox commented:
> [In Dave's comment](https://github.com/Automattic/jetpack/pull/16171#pullrequestreview-432384215) it appears between the toolbar and placeholder. Did that change?

Exactly. We had to move it in order to fix the visual reported by you. Now, it's injected in the same place that the `video` component.

<img src="https://user-images.githubusercontent.com/77539/84960914-5e050300-b0d9-11ea-854b-de39814fdc3c.png" width="400px" />

obenland commented on extensions/shared/blocks/cover/components.js:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Internal dependencies
+ */
+import UpgradeNudge from "../../components/upgrade-nudge";
```
Would you mind checking for single quotes again across the new files?

retrofox replied:
✅ I've had to re-implement it using a hook, which I think fits better to the code. 👍

retrofox commented on extensions/shared/blocks/cover/cover-media-placeholder.js:
```diff
@@ -3,86 +3,50 @@
+						<CoreMediaPlaceholder
+							{ ...props }
+							onFilesPreUpload={ onFilesUpload }
+							multiple={ false }
```
yes.

retrofox commented on extensions/shared/blocks/cover/utils.js:
```diff
@@ -1,19 +1,59 @@
+	}
+
+	// Pick up allowed mime types from the window object.
+	const allowedMimeTypes = window?.Jetpack_Editor_Initial_State?.allowedMimeTypes;
```
Yes, I've created some helper functions and put them in a single file in the shared folder in order to be usable from other code pieces.

retrofox replied:
```es6

import getAllowedMimeTypesBySite, {
	getAllowedVideoTypesByType,
	pickFileExtensionsFromMimeTypes,
} from "../../get-allowed-mime-types";

// Get all allowed mime typed defined in the site (by the server).
const allowedMimeTypes = getAllowedMimeTypesBySite();

// Get all allowed VIDEO mime types.
const allowedVideoMimeTypes = getAllowedVideoTypesByType( 'video' );

// Get allowed VIDEO file extensions .
const allowedVideoFileExtensions = pickFileExtensionsFromMimeTypes( allowedVideoMimeTypes );
```

retrofox commented on extensions/shared/blocks/cover/cover-media-placeholder.js:
```diff
@@ -3,86 +3,50 @@
+									return onFilesUpload( [ filename ] );
+								}
+								return onError( message );
+							} }
```
👍 ✅

retrofox commented:
Thanks, folks for your review. Changed and updated the code according to your feedback.
Ready for a new round. D44997-code is ✅ too.

retrofox commented:
`useBlockEditContext` needs a fallback since it isn't available in WP 5.3 triggering a fatal error.

retrofox commented:
I've updated and removed what I did (🤦 ) so we need a new ✅ here. We already did two approaches to tackle https://github.com/Automattic/jetpack/issues/16233. So the plan is to merge this one, and then choose one of the suggestions: https://github.com/Automattic/jetpack/pull/16240 or https://github.com/Automattic/jetpack/pull/16241. Both branches have been branched off from this one.

cc @jeherve @obenland 🙇‍♂️

retrofox commented:
r209519-wpcom
